### DISCUSSION TITLE
IgnoreMap alternative

### DISCUSSION THREAD

**Discussion by ArnaudB88** [LuckyPennySoftware/AutoMapper#4041] [NA]
Hi, as described in the 11.0 Upgrade guide, the IgnoreMap attribute is removed. The suggestion is to replace it with a fluent API alternative or implement the attribute ourselves.

We use the IgnoreMap attribute to exclude child entities/navigation properties of (efcore) entities from the mapping. The reason is that we want to control how child entities are added/updated/removed. Example:
```
public class OrderDto {//Source
    public Guid Id { get; set; }
    public string Name { get; set; }
    public IEnumerable<OrderLineDto> OrderLines { get; set; }
}
public class Order {//Destination
    public Guid Id { get; set; }
    public string Name { get; set; }
    [IgnoreMap]
    public IEnumerable<OrderLine> OrderLines { get; set; }
}

//Inside a Profile class
CreateMap<OrderDto, Order>(MemberList.None);
CreateMap<OrderLineDto, OrderLine>(MemberList.None);

//Inside business logic
var existingOrder = OrderRepository.GetById(orderDto.Id);//fetch from database
Mapper.Map(orderDto, existingOrder);//map all properties except the ones with IgnoreMap
//Custom logic to update the orderlines
```

Since we can have multiple FooDto's which map to a Foo entity, it's too much work to replace the IgnoreMap with the fluent api. Also when our database model is changed, all navigation properties are ignored automatically with the IgnoreMap attribute (entity classes are generated automatically with the IgnoreMap attribute on navigation properties).

My question: is there no alternative than adding the IgnoreMap attribute in our code if we want de use similar functionality (= igore a destination member once for all mappings)?

**Comment by jbogard** [comment]
There's a couple options:
- Use [attribute mapping](https://docs.automapper.org/en/stable/Attribute-mapping.html) exclusively
- Configure `ForAllMembers` to look for that `Ignore` attribute to ignore the member

The `ForAllMembers` lets you inspect the destination member as a `MemberInfo` to see if an attribute exists.

**Reply by ArnaudB88** [reply]
@jbogard Thank you for the fast answer.
About your suggestions:
- Attribute mapping exclusively is impossible due to many and complex mappings
- ForAllMembers to check Ignore attribute: do you mean to implement the suggested [IgnoreMap implementation](https://github.com/AutoMapper/AutoMapper/blob/b8f8546da66cccbe8c274358f5448feac993fee7/src/UnitTests/AutoMapperSpecBase.cs#L159) from the upgrade guide, **but reuse** the existing IgnoreAttribute instead of adding the removed IgnoreMapAttribute?

**Reply by ArnaudB88** [reply]
fyi: I implemented the second suggestion and most unit tests succeed, except mappings of some ignored properties. Those succeeded in 10.0 but are ignored in 11.0 (which makes more sense):
```
public class OrderDto {//Source
    public Guid Id { get; set; }
    public string Name { get; set; }
    public ProjectDto Project { get; set; }
}
public class Order {//Destination
    public Guid Id { get; set; }
    public string Name { get; set; }
    public Guid? ProjectId { get; set; }
    [IgnoreMap]
    public Project Project { get; set; }
}
var entity = new Order {Project = new Project()};
repository.insert(entity);
repository.SaveChanges();//Saves to database, fills entity.ProjectId with entity.Project.Id
var dto = new OrderDto{ Id = entity.Id, Project = null };//goal is to remove the order.project link again
Mapper.Map(dto, entity);//In v10 entity.ProjectId was set to null, in v11 it kept its original value
```

**Reply by jbogard** [reply]
> ForAllMembers to check Ignore attribute: do you mean to implement the suggested [IgnoreMap implementation](https://github.com/AutoMapper/AutoMapper/blob/b8f8546da66cccbe8c274358f5448feac993fee7/src/UnitTests/AutoMapperSpecBase.cs#L159) from the upgrade guide, but reuse the existing IgnoreAttribute instead of adding the removed IgnoreMapAttribute?

Yes to reuse the existing attribute, but you could also create your own `IgnoreMapAttribute`. Whatever is clearer to you and your team!
